### PR TITLE
Add basic unit tests for a lxdContainer.

### PR DESCRIPTION
**Pull request by pcdummy**
Signed-off-by: René Jochum <rene@jochums.at>

### CODE REVIEW

**Comment by pcdummy** [comment]
Working on it... forgot a stashed change.

File: Makefile [hunk 1]
```diff
@@ -29,6 +29,7 @@ protobuf:
 
 .PHONY: check
 check: default
+	go get github.com/stretchr/testify
```
Comment (stgraber) [inline Makefile]: Doesn't "go get ./..." also pull the imports of _test.go files?

File: lxd/container.go [hunk 2]
```diff
@@ -329,6 +331,15 @@ func containerLXDCreateInternal(
...
 		log.Ctx{"container": name, "id": id})
 
+	myConfig := map[string]string{}
+	if err := shared.DeepCopy(&args.Config, &myConfig); err != nil {
+		return nil, err
+	}
+	myDevices := shared.Devices{}
+	if err := shared.DeepCopy(&args.Devices, &myDevices); err != nil {
+		return nil, err
+	}
```
Comment (stgraber) [inline lxd/container.go]: How's that related to testing? ^

File: lxd/container.go [hunk 3]
```diff
@@ -166,10 +166,12 @@ func containerLXDCreateAsEmpty(d *Daemon, name string,
...
-	// Now create the empty storage
-	if err := c.Storage.ContainerCreate(c); err != nil {
-		c.Delete()
-		return nil, err
+	if !d.IsMock {
+		// Now create the empty storage
+		if err := c.Storage.ContainerCreate(c); err != nil {
+			c.Delete()
+			return nil, err
+		}
```
Comment (stgraber) [inline lxd/container.go]: Why can't the mock storage driver just not do anything on ContainerCreate(), avoiding the if statement here?

File: lxd/containers.go [hunk 4]
```diff
@@ -213,7 +213,7 @@ func containersRestart(d *Daemon) error {
 }
 
 func containersShutdown(d *Daemon) error {
-	results, err := d.ListRegularContainers()
+	results, err := dbContainersList(d.db, cTypeRegular)
```
Comment (stgraber) [inline lxd/containers.go]: How's that related to testing?

File: lxd/daemon.go [hunk 5]
```diff
@@ -470,21 +472,38 @@ SELECT fingerprint FROM images WHERE cached=1 AND last_use_date<=strftime('%s',
...
 	}
 
-	shared.Log.Info("LXD is starting")
+	if !d.IsMock {
+		shared.Log.Info("LXD is starting.",
+			log.Ctx{"path": shared.VarPath("")})
+	} else {
+		shared.Log.Info("Mock LXD is starting.",
+			log.Ctx{"path": shared.VarPath("")})
+	}
```
Comment (stgraber) [inline lxd/daemon.go]: Those strings shouldn't end with a final dot. That's a regression from the change I've pushed earlier today.

File: lxd/daemon.go [hunk 6]
```diff
@@ -632,88 +658,89 @@ func startDaemon() (*Daemon, error) {
 		NotFound.Render(w)
 	})
```
Comment (stgraber) [inline lxd/daemon.go]: So instead of re-indenting half the function, can you not let the normal socket initialization code run and simply not spawn the listeners at the end?

File: lxd/daemon.go [hunk 7]
```diff
@@ -632,88 +658,89 @@ func startDaemon() (*Daemon, error) {
...
+			tcpl, err := tls.Listen("tcp", listenAddr, tlsConfig)
+			if err != nil {
+				return fmt.Errorf("cannot listen on https socket: %v", err)
+			}
+
+			sockets = append(sockets, tcpl)
+		}
 	}
 
 	d.Sockets = sockets
```
Comment (stgraber) [inline lxd/daemon.go]: simply set sockets to empty right here when running in mock mode and that'll save you all of the above indent ^

File: lxd/daemon.go [hunk 8]
```diff
@@ -796,7 +806,10 @@ func (d *Daemon) Stop() error {
 	}
 
 	if n, err := d.numRunningContainers(); err != nil || n == 0 {
-		shared.Debugf("Unmounting shmounts: err %s n %d", err, n)
+		shared.Log.Debug(
+			"daemon.Stop: unmounting shmounts",
```
Comment (stgraber) [inline lxd/daemon.go]: Current logging standard is:
 - All strings start with a capital letter
 - No strings end with a dot
 - Function name if useful can be provided through context

**Comment by stgraber** [comment]
Stopping the review where I'm at now as I'd rather re-review once those points have been addressed, should make for a much smaller delta.

File: lxd/container.go (hunk 3 above)
Comment (pcdummy) [inline lxd/container.go]: Thats a leftover of the previous code, will remove it.

File: lxd/containers.go (hunk 4 above)
Comment (pcdummy) [inline lxd/containers.go]: Its another commit in this branch.

File: lxd/daemon.go (hunk 6 above)
Comment (pcdummy) [inline lxd/daemon.go]: Should work, want the extra function or this?

File: lxd/container.go (hunk 2 above)
Comment (pcdummy) [inline lxd/container.go]: That fixes a bug i've found during testing (related to issue #996).

File: Makefile (hunk 1 above)
Comment (pcdummy) [inline Makefile]: No it doesn't not sure its good or bad.

File: lxd/daemon.go (hunk 6 above)
Comment (stgraber) [inline lxd/daemon.go]: For now let's go with just the if statement, that should reduce the size of the diff. If at some point we need to shrink the function, we'll split it up then.

File: lxd/container.go (hunk 2 above)
Comment (stgraber) [inline lxd/container.go]: Can you split it in a separate commit? Should make it simpler to merge/rebase as @mikemccracken is working on fixes for the #996 fallout.

File: lxd/containers.go (hunk 4 above)
Comment (stgraber) [inline lxd/containers.go]: Oops, for some reason I thought I was doing commit by commit review, not the whole diff...

File: Makefile (hunk 1 above)
Comment (stgraber) [inline Makefile]: Just looked into it, we simply need to change to "go get -t ./..."

**Comment by pcdummy** [comment]
Thanks for your review @stgraber, i implemented all your suggestions above.
